pifi/videoprocessor.py

```diff
@@ -217,11 +224,21 @@ def __download_youtube_video(self):
         )
 
     def __populate_avg_color_frames(
-        self, avg_color_frames, ffmpeg_to_python_fifo, vid_start_time, bytes_per_frame, np_array_shape
+        self, avg_color_frames, ffmpeg_to_python_fifo, fps_fifo, vid_start_time, bytes_per_frame, np_array_shape
     ):
-        is_ready_to_read, ignore1, ignore2 = select.select([ffmpeg_to_python_fifo], [], [], 0)
+        select_list = [ffmpeg_to_python_fifo]
+        if not fps_fifo.closed:
+            select_list.append(fps_fifo)
+        is_ready_to_read, ignore1, ignore2 = select.select(select_list, [], [], 0)
         if not is_ready_to_read:
-            return [False, vid_start_time]
+            return [False, vid_start_time, 0]
+
+        if is_ready_to_read[0] == fps_fifo:
+            fps_parts = fps_fifo.readlines()[0]
```

Comment:
1) i think we can use `.readline()` here: https://docs.python.org/3/library/io.html#io.TextIOBase.readline

2) correct me if I'm wrong, but the way you've structured this, it looks like we could start returning data from `ffmpeg_to_python_fifo` before we've retrieved anything from the `fps_fifo`. This might happen if `ffprobe` is randomly slow. So we might start playing the video without retrieving the real FPS yet, which could lead to out of sync audio/video. 

I think it might be preferable to block playback until we retrieve the FPS, because then we avoid the out of sync issues. Although I guess this might lead to slightly longer start up times (I'm curious how much longer / if it's noticeably longer). If we do this, we can retrieve the FPS in a separate method that is called before `__populate_avg_color_frames()`, which also seems nice because then we don't have to shove FPS logic into this method. So at L180, where you currently have: `fps_fifo = open(fps_fifo_name, 'r')`, we could replace that line with something like: `fps = self.__read_fps_from_fifo(self, fps_fifo_name)`. We wouldn't need a `select.select()` thing there either because `readline()` should block until the fifo has stuff to read IIUC.

Curious what you think about that.

Reply:
>I think it might be preferable to block playback until we retrieve the FPS, because then we avoid the out of sync issues

I couldn't sleep and realized that this might not work. If we block playback until we retrieve the FPS, i.e. we don't read from the ffmpeg fifo until we retrieve the FPS, then ytdl might be blocked writing to the FPS fifo because of shell backpressure. Essentially, we have a pipeline:

```
ytdl | tee >(ffprobe fps.fifo) >(ffmpeg ffmpeg.fifo) >(ffplay ...) >/dev/null
```

If we don't read from the ffmpeg fifo, ytdl will be blocked writing to the tee, thus we may never be able to retrieve the fps. Furthermore, even if this shell backpressure scenario doesn't end up happening, ffplay may start playing audio too early. I have some ideas for workarounds:

proc 1:
```
ytdl | tee >(ffprobe fps.fifo) | mbuffer 
```

proc 2:
```
tee >(ffmpeg ffmpeg.fifo) >(ffplay ...) >/dev/null
```

If we invoke proc 1's mbuffer with a buffer size large enough for ffprobe to get the FPS without being blocked on writing to mbuffer, then once the FPS has been retrieved from the FPS fifo, we can connect the stdout of proc 1 to the stdin of proc 2.

LMK if that makes sense... Also I can help implement this if you'd prefer.